idaes/generic_models/properties/core/coolprop/coolprop_wrapper.py

```diff
@@ -0,0 +1,484 @@
+    """
+
+    def __init__(self, prop, comp):
+        msg = (f"Found unsupported expression form for {prop} "
+               f"of component {comp}. This likely occured due to "
+               f"changes in CoolProp and the interface should be "
+               f"updated.")
+
+        super().__init__(self, msg)
+
+
+class CoolPropPropertyError(KeyError):
+    """
+    Error message for when a parameter is called for that is not in CoolProp's
+    database.
+    """
+
+    def __init__(self, prop, comp):
+        msg = (f"Could not retrieve parameters for {prop} of "
+               f"component {comp} from CoolProp. This likely indicates "
+               f"that CoolProp does not have values for the necessary "
+               f"parameters.")
+
+        super().__init__(self, msg)
+
+
+class CoolPropWrapper:
+    """
+    Interface wrapper for calling CoolProp parameter database from within the
+    IDAES Generic Property Package Framework.
+
+    This class is intended to be used in theromphysical property definition
+    dicts and directs the properties framework to use forms and parameter data
+    from the CoolProp libraries (if available). This requires that the user
+    have CoolProp installed locally.
+    """
+    _cached_components = {}
+
+    @staticmethod
+    def get_parameter_value(comp_name, param):
+        """
+        Method to retrieve tuples of (value, units) for the specified property
+        and component for the critical state section of the CoolProp json
+        format.
+
+        Args:
+            comp_name - name of componet for which to retirve parameters
+            param - name of parameter to return value and units
+
+        Returns:
+            tuple of parameter (value, units)
+        """
+        try:
+            map_tuple = name_map[param]
+        except KeyError:
+            raise BurntToast(
+                "Unrecognized property name in CoolProp wrapper. Please "
+                "contact the IDAES developers with this bug.")
+
+        if map_tuple[1] == "CRITICAL":
+            ptuple = CoolPropWrapper._get_critical_property(
+                comp_name, map_tuple[0])
+        elif map_tuple[1] == "EOS":
+            ptuple = CoolPropWrapper._get_eos_property(
+                comp_name, map_tuple[0])
+        else:
+            raise BurntToast(
+                "Unrecognized argument in CoolProp wrapper. Please contact "
+                "the IDAES developers with this bug.")
+
+        return ptuple
+
+    @staticmethod
+    def flush_cached_components():
+        """
+        Method to clear cached component parameter data. This sets
+        _cached_components to an empty dict.
+
+        Args:
+            None
+
+        Returns:
+            None
+        """
+        CoolPropWrapper._cached_components = {}
+
+    # -------------------------------------------------------------------------
+    # Pure component property sub-classes
+    class dens_mol_liq_comp():
```

Comment:
I have had other reviewers argue against that.

Reply:
yeah, it's just an extra 8 characters in Python 3, and we have no intent of supporting Python 2.x

Reply:
The parens should also be dropped.

```
class dens_mol_liq_comp:
   """And awaaaay we go!
   """
```

Reply:
However, since there *is* a common base interface, you may want to make that clearer with an actual base class (that does nothing):
```
class PureComponentProperty:
  @staticmethod
  def build_parameters(cobj):
     pass

  @staticmethod
  def get_expression(*args):
    pass

```
Note that I put `get_expression` instead of `return_expression` since unless I'm missing something that is what it's doing from the caller's perspective.

Reply:
Whilst it is true that there is a common base interface, I do wonder about the value of such a bare-bones base class (which would also require some tests, trivial as they may be).

`get_expression` is a better name, but at this point don't want to break backwards compatibility as that name is used everywhere in the modular property framework.